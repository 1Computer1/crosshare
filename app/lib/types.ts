import * as t from "io-ts";
import type { WordDBT } from './WordDB';

import { DBPuzzleT, CommentWithRepliesV } from '../lib/dbtypes';

export type Optionalize<T extends K, K> = Omit<T, keyof K>;
export type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

export const BLOCK = ".";

export enum Direction {
  Across,
  Down
}

export interface WorkerMessage {
  type: string,
}
export interface AutofillResultMessage extends WorkerMessage {
  type: 'autofill-result',
  input: string[],
  result: string[]
}
export function isAutofillResultMessage(msg: WorkerMessage): msg is AutofillResultMessage {
  return msg.type === 'autofill-result';
}
export interface AutofillCompleteMessage extends WorkerMessage {
  type: 'autofill-complete',
}
export function isAutofillCompleteMessage(msg: WorkerMessage): msg is AutofillCompleteMessage {
  return msg.type === 'autofill-complete';
}
export interface LoadDBMessage extends WorkerMessage {
  type: 'loaddb',
  db: WordDBT,
}
export function isLoadDBMessage(msg: WorkerMessage): msg is LoadDBMessage {
  return msg.type === 'loaddb';
}
export interface AutofillMessage extends WorkerMessage {
  type: 'autofill',
  grid: string[],
  width: number,
  height: number
}
export function isAutofillMessage(msg: WorkerMessage): msg is AutofillMessage {
  return msg.type === 'autofill';
}

export interface Position {
  row: number,
  col: number,
}

export interface PosAndDir extends Position {
  dir: Direction,
}

const PuzzleJsonMandatoryV = t.type({
  title: t.string,
  size: t.type({
    rows: t.number,
    cols: t.number
  }),
  clues: t.type({
    across: t.array(t.string),
    down: t.array(t.string)
  }),
  grid: t.array(t.string)
});
const PuzzleJsonOptionalV = t.partial({
  highlighted: t.array(t.number),
  highlight: t.keyof({ circle: null, shade: null }),
});
export const PuzzleJsonV = t.intersection([PuzzleJsonMandatoryV, PuzzleJsonOptionalV]);
export type PuzzleJson = t.TypeOf<typeof PuzzleJsonV>;

const ClueV = t.type({
  num: t.number,
  dir: t.union([t.literal(0), t.literal(1)]),
  clue: t.string,
});
export type ClueT = t.TypeOf<typeof ClueV>;

export const PuzzleV = t.type({
  authorId: t.string,
  category: t.union([t.string, t.null]),
  authorName: t.string,
  moderated: t.boolean,
  publishTime: t.union([t.number, t.null]),
  title: t.string,
  size: t.type({
    rows: t.number,
    cols: t.number
  }),
  clues: t.array(ClueV),
  grid: t.array(t.string),
  highlighted: t.array(t.number),
  highlight: t.keyof({ circle: null, shade: null }),
  comments: t.array(CommentWithRepliesV),
});

export type PuzzleT = t.TypeOf<typeof PuzzleV>;
export type PuzzleResult = PuzzleT & { id: string };

export function puzzleTitle(puzzle: PuzzleT) {
  let title = puzzle.title;
  if (puzzle.category === 'dailymini' && puzzle.publishTime) {
    const d = new Date(puzzle.publishTime);
    const ds = (d.getUTCMonth() + 1) + "/" + d.getUTCDate() + "/" + d.getUTCFullYear()
    title = "Daily Mini for " + ds;
  }
  return title;
}

export function puzzleFromDB(dbPuzzle: DBPuzzleT): PuzzleT {
  let clues: Array<ClueT> = [];
  for (let i = 0; i < dbPuzzle.ac.length; i += 1) {
    clues.push({ dir: Direction.Across, clue: dbPuzzle.ac[i], num: dbPuzzle.an[i] });
  }
  for (let i = 0; i < dbPuzzle.dc.length; i += 1) {
    clues.push({ dir: Direction.Down, clue: dbPuzzle.dc[i], num: dbPuzzle.dn[i] });
  }
  return {
    authorId: dbPuzzle.a,
    category: dbPuzzle.c,
    authorName: dbPuzzle.n,
    moderated: dbPuzzle.m,
    publishTime: dbPuzzle.p ? dbPuzzle.p.toMillis() : null,
    title: dbPuzzle.t,
    size: {
      rows: dbPuzzle.h,
      cols: dbPuzzle.w
    },
    clues: clues,
    grid: dbPuzzle.g,
    highlighted: dbPuzzle.hs || [],
    highlight: dbPuzzle.s ? "shade" : "circle",
    comments: dbPuzzle.cs || []
  };
}
