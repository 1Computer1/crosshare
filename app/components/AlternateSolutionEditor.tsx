import useEventListener from '@use-it/event-listener';
import { useRef, useCallback, useReducer, useMemo } from 'react';
import { FaSave, FaWindowClose, FaEllipsisH, FaVolumeUp, FaVolumeMute, FaKeyboard } from 'react-icons/fa';
import { usePersistedBoolean } from '../lib/hooks';
import { Direction, fromKeyboardEvent, fromKeyString, KeyK } from '../lib/types';
import { fromCells } from '../lib/viewableGrid';
import { Rebus, EscapeKey } from './Icons';
import { Keyboard } from './Keyboard';
import { TopBar, TopBarLink, TopBarDropDown, TopBarDropDownLink } from './TopBar';
import { gridInterfaceReducer, KeypressAction, PasteAction } from '../reducers/reducer';
import { Square } from './Square';
import { GridView } from './Grid';
import { isSome } from 'fp-ts/lib/Option';

export function AlternateSolutionEditor(props: { grid: string[], width: number, height: number, highlighted: Set<number>, highlight: 'circle' | 'shade', cancel: () => void, save: (alt: Record<number, string>) => Promise<void> }) {
  const initialGrid = fromCells({
    mapper: (e) => e,
    width: props.width,
    height: props.height,
    cells: props.grid,
    allowBlockEditing: false,
    highlighted: props.highlighted,
    highlight: props.highlight,
  });

  const [state, dispatch] = useReducer(gridInterfaceReducer, {
    type: 'alternate-editor',
    active: { col: 0, row: 0, dir: Direction.Across },
    grid: initialGrid,
    wasEntryClick: false,
    showExtraKeyLayout: false,
    isEnteringRebus: false,
    rebusValue: '',
    downsOnly: false,
    isEditable: () => true,
  });

  const [muted, setMuted] = usePersistedBoolean('muted', false);
  const [toggleKeyboard, setToggleKeyboard] = usePersistedBoolean(
    'keyboard',
    false
  );

  const gridRef = useRef<HTMLDivElement | null>(null);

  const focusGrid = useCallback(() => {
    if (gridRef.current) {
      gridRef.current.focus();
    }
  }, []);

  const physicalKeyboardHandler = useCallback(
    (e: KeyboardEvent) => {
      const tagName = (e.target as HTMLElement)?.tagName?.toLowerCase();
      if (tagName === 'textarea' || tagName === 'input') {
        return;
      }
      const mkey = fromKeyboardEvent(e);
      if (isSome(mkey)) {
        const kpa: KeypressAction = { type: 'KEYPRESS', key: mkey.value };
        dispatch(kpa);
        e.preventDefault();
      }
    },
    [dispatch]
  );
  useEventListener(
    'keydown',
    physicalKeyboardHandler,
    gridRef.current || undefined
  );

  const pasteHandler = useCallback((e: ClipboardEvent) => {
    const tagName = (e.target as HTMLElement)?.tagName?.toLowerCase();
    if (tagName === 'textarea' || tagName === 'input') {
      return;
    }
    const pa: PasteAction = {
      type: 'PASTE',
      content: e.clipboardData?.getData('Text') || ''
    };
    dispatch(pa);
    e.preventDefault();
  }, [dispatch]);
  useEventListener(
    'paste',
    pasteHandler
  );

  const keyboardHandler = useCallback(
    (key: string) => {
      const mkey = fromKeyString(key);
      if (isSome(mkey)) {
        const kpa: KeypressAction = { type: 'KEYPRESS', key: mkey.value };
        dispatch(kpa);
      }
    },
    [dispatch]
  );

  const topBarChildren = useMemo(() => {
    return (
      <>
        <TopBarLink
          icon={<FaSave />}
          text="Add Alternate"
          onClick={async () => {
            const alt: Record<number, string> = {};
            let hadAny = false;
            for (const [idx, cellValue] of state.grid.cells.entries()) {
              const defaultCellValue = initialGrid.cells[idx];
              if (cellValue.trim() && cellValue.trim() != defaultCellValue?.trim()) {
                hadAny = true;
                alt[idx] = cellValue;
              }
            }
            if (!hadAny) {
              props.cancel();
              return;
            }
            return props.save(alt).then(() => props.cancel());
          }}
        />
        <TopBarLink
          icon={<FaWindowClose />}
          text="Cancel"
          onClick={props.cancel}
        />
        <TopBarDropDown onClose={focusGrid} icon={<FaEllipsisH />} text="More">
          {() => (
            <>
              <TopBarDropDownLink
                icon={<Rebus />}
                text="Enter Rebus"
                shortcutHint={<EscapeKey />}
                onClick={() => {
                  const a: KeypressAction = {
                    type: 'KEYPRESS',
                    key: { k: KeyK.Escape },
                  };
                  dispatch(a);
                }}
              />
              {muted ? (
                <TopBarDropDownLink
                  icon={<FaVolumeUp />}
                  text="Unmute"
                  onClick={() => setMuted(false)}
                />
              ) : (
                <TopBarDropDownLink
                  icon={<FaVolumeMute />}
                  text="Mute"
                  onClick={() => setMuted(true)}
                />
              )}
              <TopBarDropDownLink
                icon={<FaKeyboard />}
                text="Toggle Keyboard"
                onClick={() => setToggleKeyboard(!toggleKeyboard)}
              />
            </>
          )}
        </TopBarDropDown>
      </>
    );
  }, [focusGrid, initialGrid.cells, muted, props, setMuted, setToggleKeyboard, state.grid.cells, toggleKeyboard]);

  const parentRef = useRef<HTMLDivElement | null>(null);

  return (
    <>
      <div
        css={{
          display: 'flex',
          flexDirection: 'column',
          height: '100%',
        }}
      >
        <div css={{ flex: 'none' }}>
          <TopBar>{topBarChildren}</TopBar>
        </div>
        <div
          css={{ flex: '1 1 auto', overflow: 'scroll', position: 'relative' }}
        >
          <div
            // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
            tabIndex={0}
            ref={(instance) => {
              parentRef.current = instance;
            }}
            css={{
              outline: 'none',
              display: 'flex',
              flex: '1 1 auto',
              flexDirection: 'column',
              alignItems: 'center',
              height: '100%',
              width: '100%',
              position: 'absolute',
              flexWrap: 'nowrap',
            }}
          >
            <Square
              noColumns={true}
              waitToResize={false}
              parentRef={parentRef}
              aspectRatio={props.width / props.height}
              contents={(width: number, _height: number) => {
                return (
                  <GridView
                    isEnteringRebus={state.isEnteringRebus}
                    rebusValue={state.rebusValue}
                    squareWidth={width}
                    grid={state.grid}
                    defaultGrid={initialGrid}
                    active={state.active}
                    dispatch={dispatch}
                    allowBlockEditing={false}
                    autofill={[]}
                  />
                );
              }}
            />
          </div>
        </div>
        <div css={{ flex: 'none', width: '100%' }}>
          <Keyboard
            toggleKeyboard={toggleKeyboard}
            keyboardHandler={keyboardHandler}
            muted={muted}
            showExtraKeyLayout={state.showExtraKeyLayout}
            includeBlockKey={false}
          />
        </div>
      </div>
    </>
  );
}