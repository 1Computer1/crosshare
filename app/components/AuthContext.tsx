import { ReactNode, createContext, useContext } from 'react';
import { ErrorPage } from '../components/ErrorPage';
import type firebase from 'firebase/app';
import { GoogleLinkButton, GoogleSignInButton } from './GoogleButtons';
import { TopBar } from './TopBar';
import { ConstructorPageT } from '../lib/constructorPage';
import { NotificationT } from '../lib/notifications';
import { AccountPrefsT } from '../lib/prefs';
import { WithConditionalCSSProp } from '@emotion/react/types/jsx-namespace';

export interface AuthProps {
  isAdmin: boolean;
  user: firebase.User;
  constructorPage?: ConstructorPageT;
  prefs?: AccountPrefsT;
}

export interface AuthPropsOptional {
  isAdmin: boolean;
  user?: firebase.User;
  constructorPage?: ConstructorPageT;
  prefs?: AccountPrefsT;
}

export function renderLoginButtonIfNeeded({
  user,
  loading,
  error,
}: AuthContextValue): React.ReactNode | null {
  if (loading) {
    return <div></div>;
  }
  if (error) {
    return <div>Error loading user: {error}</div>;
  }
  if (!user) {
    return <GoogleSignInButton />;
  }
  if (user.isAnonymous) {
    return <GoogleLinkButton user={user} />;
  }
  return null;
}

function renderLoginIfNeeded({
  user,
  loading,
  error,
}: AuthContextValue): React.ReactNode | null {
  if (loading) {
    return <div></div>;
  }
  if (error) {
    return <div>Error loading user: {error}</div>;
  }
  if (!user) {
    return (
      <>
        <TopBar />
        <div css={{ margin: '1em' }}>
          <p>
            Please sign-in with your Google account to continue. We use your
            account to keep track of the puzzles you&apos;ve played and your
            solve streaks.
          </p>
          <GoogleSignInButton />
        </div>
      </>
    );
  }
  if (user.isAnonymous) {
    return (
      <>
        <TopBar />
        <div css={{ margin: '1em' }}>
          <p>
            Please sign-in with your Google account to continue. We use your
            account to keep track of the puzzles you&apos;ve played and your
            solve streaks.
          </p>
          <GoogleLinkButton user={user} />
        </div>
      </>
    );
  }
  return null;
}

/* Ensure we have a non-anonymous user, upgrading an anonymous user if we have one. */
export function requiresAuth<T>(WrappedComponent: React.ComponentType<T>) {
  return (props: T & WithConditionalCSSProp<T>): ReactNode => {
    const ctx = useContext(AuthContext);
    const login = renderLoginIfNeeded(ctx);
    if (login) {
      return login;
    }
    return (
      <WrappedComponent
        {...props}
        isAdmin={ctx.isAdmin}
        user={ctx.user}
        constructorPage={ctx.constructorPage}
        prefs={ctx.prefs}
      />
    );
  };
}

/* Ensure we have an admin user, upgrading an anonymous user if we have one. */
export function requiresAdmin<T>(WrappedComponent: React.ComponentType<T>) {
  return (props: T & WithConditionalCSSProp<T>): ReactNode => {
    const ctx = useContext(AuthContext);
    const login = renderLoginIfNeeded(ctx);
    if (login) {
      return login;
    }
    if (!ctx.isAdmin) {
      return (
        <ErrorPage title="Not Allowed">
          <p>You do not have permission to view this page</p>
        </ErrorPage>
      );
    }
    return (
      <WrappedComponent
        {...props}
        isAdmin={true}
        user={ctx.user}
        constructorPage={ctx.constructorPage}
        prefs={ctx.prefs}
      />
    );
  };
}

interface AuthContextValue {
  user?: firebase.User;
  notifications?: Array<NotificationT>;
  isAdmin: boolean;
  loading: boolean;
  error?: string;
  constructorPage?: ConstructorPageT;
  prefs?: AccountPrefsT;
}
export const AuthContext = createContext({
  user: undefined,
  loading: false,
  error: 'using default context',
  constructorPage: undefined,
  prefs: undefined,
} as AuthContextValue);
