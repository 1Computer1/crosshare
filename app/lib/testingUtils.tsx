import { ReactElement, ComponentType } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { matchers, createSerializer } from 'jest-emotion';

import { AuthContext } from '../components/AuthContext';

import type firebaseTypes from 'firebase';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers);
expect.addSnapshotSerializer(createSerializer());

export const anonymousUser = {
  uid: 'anonymous-user-id',
  isAnonymous: true
} as firebaseTypes.User;

const WithAllProviders: ComponentType = ({ children }) => {
  return (
    <AuthContext.Provider value={{ user: anonymousUser, isAdmin: false, loadingUser: false, error: undefined }}>
      {children}
    </AuthContext.Provider>
  )
}

function wrappedRender(
  ui: ReactElement,
  options?: RenderOptions
): RenderResult {

  return render(ui, {
    wrapper: WithAllProviders,
    ...options,
  });
}
// re-export everything
export * from '@testing-library/react';

// override render method
export { wrappedRender as render };
