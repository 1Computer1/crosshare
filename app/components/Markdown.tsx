/* eslint-disable @typescript-eslint/no-explicit-any */

import { ReactNode, Fragment } from 'react';
import SimpleMarkdown from 'simple-markdown';
import { useHover } from '../lib/hooks';
import { Direction } from '../lib/types';
import { ToolTipText } from './ToolTipText';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const { image, refimage, ...baseRules } = { ...SimpleMarkdown.defaultRules };

const rules: SimpleMarkdown.Rules<SimpleMarkdown.ReactOutputRule> = {
  ...baseRules,
  blockQuote: {
    ...SimpleMarkdown.defaultRules.blockQuote,
    match: SimpleMarkdown.blockRegex(/^( *>[^!\n]+(\n[^\n]+)*\n*)+\n{2,}/),
  },
  link: {
    ...SimpleMarkdown.defaultRules.link,
    react(node: any, output: any, state: any) {
      return SimpleMarkdown.reactElement(
        'a',
        state.key,
        {
          href: SimpleMarkdown.sanitizeUrl(node.target),
          target: '_blank',
          rel: 'nofollow ugc noopener noreferrer',
          title: node.title,
          children: output(node.content, state)
        }
      );
    },
  },
  spoiler: {
    order: SimpleMarkdown.defaultRules.em.order - 0.5,
    match(source: any, state: any) {
      if (state.inSpoiler) {
        return null;
      }
      return /^>!(.+?)!</.exec(source) || /^\|\|(.+?)\|\|/.exec(source);
    },
    parse(capture: any, recurseParse: any, state: any) {
      return {
        content: recurseParse(capture[1], { ...state, inSpoiler: true })
      };
    },
    react(node: any, recurseOutput: any, state: any) {
      return <SpoilerText key={state.key}>{recurseOutput(node.content, state)}</SpoilerText>;
    },
  }
};

const parser = SimpleMarkdown.parserFor(rules);
const output = SimpleMarkdown.outputFor(rules, 'react');

const SpoilerText = ({ children }: { children: ReactNode }) => {
  const [isHovered, hoverBind] = useHover();

  return <span css={{
    backgroundColor: isHovered ? 'var(--bg)' : 'var(--text)',
  }} {...hoverBind}>{children}</span>;
};

export const Markdown = ({ text, clueMap, inline }: { text: string, clueMap?: Map<string, [number, Direction, string]>, inline?: boolean }) => {
  if (clueMap && clueMap.size) {
    const regex = '^([^0-9A-Za-z\\s\\u00c0-\\uffff]*[0-9A-Za-z\\s\\u00c0-\\uffff]*)\\b(' + Array.from(clueMap.keys()).join('|') + ')\\b';
    const re = new RegExp(regex);
    const newRules = {
      ...rules,
      cluedWord: {
        order: SimpleMarkdown.defaultRules.em.order - 0.3,
        match(source: any) {
          return re.exec(source);
        },
        parse(capture: any, recurseParse: any, state: any) {
          return {
            pre: recurseParse(capture[1], state),
            content: capture[2]
          };
        },
        react(node: any, recurseOutput: any, state: any) {
          const mouseover = clueMap.get(node.content);
          if (!mouseover) {
            throw new Error('expected to find clue ' + node.content);
          }
          return <Fragment key={state.key}>
            {recurseOutput(node.pre, state)}
            <ToolTipText text={node.content} tooltip={
              <>
                <b css={{ marginRight: '0.5em' }}>{mouseover[0]}{mouseover[1] === Direction.Down ? 'D' : 'A'}</b>
                {mouseover[2]}
              </>
            } />
          </Fragment>;
        },
      }
    };
    return SimpleMarkdown.outputFor(newRules, 'react')(SimpleMarkdown.parserFor(newRules)(text + '\n\n', { inline }));
  }
  return output(parser(text + '\n\n', { inline }));
};
