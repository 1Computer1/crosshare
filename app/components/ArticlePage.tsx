import Head from 'next/head';
import { ArticleT } from '../lib/article';
import { ArticlePageProps } from '../lib/serverOnly';
import { HUGE_AND_UP, MAX_WIDTH } from '../lib/style';
import { ErrorPage } from './ErrorPage';
import { Link } from './Link';
import { Markdown } from './Markdown';
import { DefaultTopBar } from './TopBar';

export function ArticlePage(props: ArticlePageProps) {
  if ('error' in props) {
    return (
      <ErrorPage title="Article Not Found">
        <p>
          We&apos;re sorry, we couldn&apos;t find the article you requested.
        </p>
        <p>{props.error}</p>
        <p>
          Try the <Link href="/">homepage</Link>.
        </p>
      </ErrorPage>
    );
  }
  return <Article key={props.s} {...props} />;
}

function Article(props: ArticleT) {
  return (
    <>
      <Head>
        <title>{props.t}</title>
        <meta key="og:title" property="og:title" content={props.t} />
        <link
          rel="canonical"
          href={'https://crosshare.org/article/' + props.s}
        />
      </Head>
      <DefaultTopBar />
      <Markdown
        text={props.c}
        css={{
          margin: '2em 1em',
          [HUGE_AND_UP]: {
            maxWidth: MAX_WIDTH,
            margin: '2em auto',
          },
        }}
      />
    </>
  );
}
