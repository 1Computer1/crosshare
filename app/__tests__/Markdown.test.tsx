import { render } from '../lib/testingUtils';

import { Markdown } from '../components/Markdown';
import { Direction } from '../lib/types';

test('spoiler text rendering', () => {
  let r = render(<Markdown text='foo bar >!baz' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown text='foo bar >!baz!<' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown text='>!baz foo bam ! >> fooey!<' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown text='>!baz foo bam ! >> fooey!< with after text' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown text='before ||baz foo bam >! fooey|| with after text' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown text='before >!baz foo bam || fooey!< with after text' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown text='before >!baz foo bam || fooey!< with ||after|| text' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown text='before ||baz foo bam >! not! !< fooey|| with >!after!< text' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown text={'before ||baz foo bam \n\n>! not! !< fooey|| with >!after!< text'} />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown text={'before baz foo bam \n\n>! not! !< fooey|| with >!after!< text'} />, {});
  expect(r.container).toMatchSnapshot();
});

test('clueMap rendering', () => {
  const clueMap = new Map<string, [number, Direction, string]>([
    ['BAM', [2, 1, 'here is the clue?']],
    ['12ACLUE1', [45, 0, 'Well now']],
  ]);

  let r = render(<Markdown text='before ||baz BOOM foo BAM >! not! !< fooey|| with >!after!< text' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown clueMap={clueMap} text='before ||baz BOOM foo BAM >! not! !< fooey|| with >!after!< text' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown clueMap={clueMap} text='12ACLUE1 BAM' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown clueMap={clueMap} text='||BAM||' />, {});
  expect(r.container).toMatchSnapshot();

  r = render(<Markdown clueMap={clueMap} text={'You got it!! Glad the clues pointed you in the right direction. That\'s what they\'re there for. Also, it was Brian\'s suggestion to include >! BAM !< which I think is such an awesome addition. Cheers!'} />, {});
  expect(r.container).toMatchSnapshot();
});
