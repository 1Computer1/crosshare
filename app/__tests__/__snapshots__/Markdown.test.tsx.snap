// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clueMap rendering 1`] = `
.emotion-0 {
  background-color: var(--text);
}

<div>
  <div
    class="paragraph"
  >
    before 
    <span
      class="emotion-0"
    >
      baz BOOM foo BAM &gt;! not! !&lt; fooey
    </span>
     with 
    <span
      class="emotion-0"
    >
      after
    </span>
     text
  </div>
</div>
`;

exports[`clueMap rendering 2`] = `
.emotion-4 {
  background-color: var(--text);
}

.emotion-0 {
  border-bottom: 1px dotted;
}

.emotion-3 {
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  padding: 10px;
  visibility: hidden;
}

.emotion-1 {
  margin-right: 0.5em;
}

.emotion-2 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-2 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-2 {
  bottom: -5px;
}

.emotion-2::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div
    class="paragraph"
  >
    before 
    <span
      class="emotion-4"
    >
      baz BOOM foo 
      <span
        class="emotion-0"
      >
        BAM
      </span>
      <div
        class="emotion-3"
        style="position: absolute; left: 0px; top: 0px;"
      >
        <b
          class="emotion-1"
        >
          2
          D
        </b>
        here is the clue?
        <div
          class="emotion-2"
        />
      </div>
       &gt;! not! !&lt; fooey
    </span>
     with 
    <span
      class="emotion-4"
    >
      after
    </span>
     text
  </div>
</div>
`;

exports[`clueMap rendering 3`] = `
.emotion-0 {
  border-bottom: 1px dotted;
}

.emotion-3 {
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  padding: 10px;
  visibility: hidden;
}

.emotion-1 {
  margin-right: 0.5em;
}

.emotion-2 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-2 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-2 {
  bottom: -5px;
}

.emotion-2::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div
    class="paragraph"
  >
    <span
      class="emotion-0"
    >
      12ACLUE1
    </span>
    <div
      class="emotion-3"
      style="position: absolute; left: 0px; top: 0px;"
    >
      <b
        class="emotion-1"
      >
        45
        A
      </b>
      Well now
      <div
        class="emotion-2"
      />
    </div>
     
    <span
      class="emotion-0"
    >
      BAM
    </span>
    <div
      class="emotion-3"
      style="position: absolute; left: 0px; top: 0px;"
    >
      <b
        class="emotion-1"
      >
        2
        D
      </b>
      here is the clue?
      <div
        class="emotion-2"
      />
    </div>
  </div>
</div>
`;

exports[`clueMap rendering 4`] = `
.emotion-4 {
  background-color: var(--text);
}

.emotion-0 {
  border-bottom: 1px dotted;
}

.emotion-3 {
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  padding: 10px;
  visibility: hidden;
}

.emotion-1 {
  margin-right: 0.5em;
}

.emotion-2 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-2 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-2 {
  bottom: -5px;
}

.emotion-2::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div
    class="paragraph"
  >
    <span
      class="emotion-4"
    >
      <span
        class="emotion-0"
      >
        BAM
      </span>
      <div
        class="emotion-3"
        style="position: absolute; left: 0px; top: 0px;"
      >
        <b
          class="emotion-1"
        >
          2
          D
        </b>
        here is the clue?
        <div
          class="emotion-2"
        />
      </div>
    </span>
  </div>
</div>
`;

exports[`clueMap rendering 5`] = `
.emotion-4 {
  background-color: var(--text);
}

.emotion-0 {
  border-bottom: 1px dotted;
}

.emotion-3 {
  border-radius: 5px;
  background-color: var(--black);
  color: var(--white);
  text-align: center;
  padding: 10px;
  visibility: hidden;
}

.emotion-1 {
  margin-right: 0.5em;
}

.emotion-2 {
  position: absolute;
  width: 10px;
  height: 10px;
}

[data-popper-placement^="bottom"] .emotion-2 {
  top: -5px;
}

[data-popper-placement^="top"] .emotion-2 {
  bottom: -5px;
}

.emotion-2::after {
  content: " ";
  position: absolute;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  width: 10px;
  height: 10px;
  background-color: var(--black);
}

<div>
  <div
    class="paragraph"
  >
    You got it!! Glad the clues pointed you in the right direction. That's what they're there for. Also, it was Brian's suggestion to include 
    <span
      class="emotion-4"
    >
       
      <span
        class="emotion-0"
      >
        BAM
      </span>
      <div
        class="emotion-3"
        style="position: absolute; left: 0px; top: 0px;"
      >
        <b
          class="emotion-1"
        >
          2
          D
        </b>
        here is the clue?
        <div
          class="emotion-2"
        />
      </div>
       
    </span>
     which I think is such an awesome addition. Cheers!
  </div>
</div>
`;

exports[`spoiler text rendering 1`] = `
<div>
  <div
    class="paragraph"
  >
    foo bar &gt;!baz
  </div>
</div>
`;

exports[`spoiler text rendering 2`] = `
.emotion-0 {
  background-color: var(--text);
}

<div>
  <div
    class="paragraph"
  >
    foo bar 
    <span
      class="emotion-0"
    >
      baz
    </span>
  </div>
</div>
`;

exports[`spoiler text rendering 3`] = `
.emotion-0 {
  background-color: var(--text);
}

<div>
  <div
    class="paragraph"
  >
    <span
      class="emotion-0"
    >
      baz foo bam ! &gt;&gt; fooey
    </span>
  </div>
</div>
`;

exports[`spoiler text rendering 4`] = `
.emotion-0 {
  background-color: var(--text);
}

<div>
  <div
    class="paragraph"
  >
    <span
      class="emotion-0"
    >
      baz foo bam ! &gt;&gt; fooey
    </span>
     with after text
  </div>
</div>
`;

exports[`spoiler text rendering 5`] = `
.emotion-0 {
  background-color: var(--text);
}

<div>
  <div
    class="paragraph"
  >
    before 
    <span
      class="emotion-0"
    >
      baz foo bam &gt;! fooey
    </span>
     with after text
  </div>
</div>
`;

exports[`spoiler text rendering 6`] = `
.emotion-0 {
  background-color: var(--text);
}

<div>
  <div
    class="paragraph"
  >
    before 
    <span
      class="emotion-0"
    >
      baz foo bam || fooey
    </span>
     with after text
  </div>
</div>
`;

exports[`spoiler text rendering 7`] = `
.emotion-0 {
  background-color: var(--text);
}

<div>
  <div
    class="paragraph"
  >
    before 
    <span
      class="emotion-0"
    >
      baz foo bam || fooey
    </span>
     with 
    <span
      class="emotion-0"
    >
      after
    </span>
     text
  </div>
</div>
`;

exports[`spoiler text rendering 8`] = `
.emotion-0 {
  background-color: var(--text);
}

<div>
  <div
    class="paragraph"
  >
    before 
    <span
      class="emotion-0"
    >
      baz foo bam &gt;! not! !&lt; fooey
    </span>
     with 
    <span
      class="emotion-0"
    >
      after
    </span>
     text
  </div>
</div>
`;

exports[`spoiler text rendering 9`] = `
.emotion-0 {
  background-color: var(--text);
}

<div>
  <div
    class="paragraph"
  >
    before ||baz foo bam 
  </div>
  <div
    class="paragraph"
  >
    <span
      class="emotion-0"
    >
       not! 
    </span>
     fooey|| with 
    <span
      class="emotion-0"
    >
      after
    </span>
     text
  </div>
</div>
`;

exports[`spoiler text rendering 10`] = `
.emotion-0 {
  background-color: var(--text);
}

<div>
  <div
    class="paragraph"
  >
    before baz foo bam 
  </div>
  <div
    class="paragraph"
  >
    <span
      class="emotion-0"
    >
       not! 
    </span>
     fooey|| with 
    <span
      class="emotion-0"
    >
      after
    </span>
     text
  </div>
</div>
`;
